celltype_annotation/analysis/marker-gene-celltype-exploration.Rmd

```diff
@@ -0,0 +1,268 @@
+Then to assign cell type labels, the scores for all cells in a cluster are summed together and the label with the top score is chosen. 
+This means that one label is assigned to all scores in a cluster. 
+To assign `NA` cells, it appears that they take any cells with a score that's in the lower 25% of the scores, but that part is more unclear. 
+I followed the guidelines in the [README of the Github page](https://github.com/IanevskiAleksandr/sc-type) and created the function below to annotate celltypes provided a reference model.
+
+```{r}
+# function for running sctype and creating a results table
+run_sctype <- function(annotated_sce,
+                       model){
+  
+  # grab the unique celltypes
+  celltypes <- model$labels$unique
+  
+  # get a list of all markers for each celltype
+  all_markers <- purrr::map(celltypes, 
+                            \(label){
+                              
+                              # grab the markers for each celltype
+                              markers <- unique(unlist(model$markers$full[[label]]))
+                              
+                            }) |>
+    purrr::set_names(celltypes)
+  
+  # get the assigned score for each cell for each celltype
+  celltype_mtx <- sctype_score(scRNAseqData = as.matrix(logcounts(annotated_sce)),
+                               scaled = TRUE,
+                               gs = all_markers,
+                               gs2 = NULL
+                               
+  )
+  
+  # assign clusters for the sce object 
+  pcs <- reducedDim(annotated_sce, "PCA")
+  annotated_sce$clusters <- bluster::clusterRows(pcs,
+                                                 bluster::NNGraphParam(cluster.fun = "louvain",
+                                                                      type = "jaccard"))
+  
+  # extract the barcode and clusters information from coldata 
+  coldata_df <- as.data.frame(colData(annotated_sce)) |>
+    tibble::rownames_to_column("barcode") |>
+    dplyr::select(barcode, clusters)
+  
+  # make a matrix of all the scores for all barcodes, clusters, and label combinations
+  score_mtx <- celltype_mtx |>
+    t() |>
+    as.data.frame() |>
+    tibble::rownames_to_column("barcode") |>
+    dplyr::left_join(coldata_df) |>
+    tidyr::pivot_longer(cols = !c("barcode", "clusters"),
+                        names_to = "celltype",
+                        values_to = "score")
+  
+  # summarize the scores by clusters and celltype so that all scores are summed across all barcodes
+  # for each label in a cluster
+  summarized_scores_mtx <- score_mtx |>
+    dplyr::group_by(clusters, celltype) |>
+    dplyr::summarise(sum_score = sum(score))
+  
+  # assign labels for each cluster, final number of rows should be equal to the number of clusters
+  # obtain one celltype label per cluster
+  cluster_celltypes <- summarized_scores_mtx |>
+    dplyr::group_by(clusters) |>
+    dplyr::add_count() |>
+    dplyr::top_n(n = 1, wt = sum_score)
```

Comment:
The matrix that is returned is a cell by label matrix so that you get a score for every cell for every label. Originally I looked at just assigning the label on a per cell basis by taking the label associated with the max score for each cell. I received similar results, so then tried the cluster method that was suggested. I can do both, but I wasn't quite sure how to assign `NA` for the per cell version since its based on the total number of cells per cluster. Maybe we use total number of cells in the dataset?

Reply:
I think the score cutoff they are using is essentially `0.25*ncells` for a score that is the sum of the scores of individual cells, so maybe just using a score cutoff for a single cell of 0.25?